### DISCUSSION TITLE
How to implement drag after long press?

### DISCUSSION THREAD

**Discussion by laptou** [software-mansion/react-native-gesture-handler#1826] [Q&A] [answered]
The docs say:

> Simply set `manualActivation` to true on a PanGesture and use StateManager to fail the gesture if the user attempts to drag the component sooner than the duration of the long press.

However, I can't figure out how to measure timing in the callbacks of the `PanGesture`, because the app crashes if I try to use `setTimeout`, and even if I were able to use `setTimeout`, I can't get a reference to the `GestureStateManager` except in the touch callbacks, so I'm not sure how to move the gesture into the `START` state.

Here's what I have so far:
```tsx

  const isPressed = useSharedValue(false);
  const isDragging = useSharedValue(false);

  const start = useSharedValue({
    x: 0,
    y: 0,
  });

  const offset = useSharedValue({
    x: 0,
    y: 0,
  });

const gesture =
  Gesture.Pan()
    .manualActivation(true)
    .onBegin((evt) => {
      console.log('pan begin');
    })
    .onStart(() => {
      console.log('pan start');
      isPressed.value = true;
      offset.value = {
        x: 0,
        y: 0,
      };
    })
    .onTouchesDown((evt, state) => {
      if (evt.numberOfTouches !== 1) {
        state.fail();
      }

      isPressed.value = true;
      start.value = {
        x: evt.allTouches[0].x,
        y: evt.allTouches[0].y,
      };

      // using setTimeout here causes a crash, and using runOnJS doesn't fix it 

      // runOnJS(setTimeout)(() => {
      //   isDragging.value = true;
      //   state.activate();
      // }, 500);
    })
    .onTouchesMove((evt, state) => {
      isPressed.value = true;

      const offsetX = start.value.x - evt.allTouches[0].x;
      const offsetY = start.value.y - evt.allTouches[0].y;

      const dist = Math.sqrt(offsetX * offsetX + offsetY * offsetY);

      if (dist > 10) {
        state.fail();
      }
    })
    .onUpdate((evt) => {
      offset.value = {
        x: evt.translationX,
        y: evt.translationY,
      };
    })
    .onFinalize(() => {
      offset.value = {
        x: 0,
        y: 0,
      };
      isPressed.value = false;
      isDragging.value = false;

      console.log('pan finalize');
    });
```

**Comment by JasonCareter** [comment, marked as answer]
well, you're supposed to use composition gestures . Passed a Gesture.Exclusive( longPressGesture, panGesture ) to the target View .
Make a SharedValue to record the state whether long press activated . So , you can activate the pan gesture manually inside the onTouchesMove callback when the long-press state is true . With this little change , you can get what you want as your question described .

**Reply by lsps9150414** [reply]
Hey @yaaliuzhipeng is it possible to provide an example snippet? I've tried but it didn't work, not sure if I'm doing it right.

**Reply by JasonCareter** [reply]
> Hey @yaaliuzhipeng is it possible to provide an example snippet? I've tried but it didn't work, not sure if I'm doing it right.

Are u Chinese? Sure , I’m kind of busy right now and I will post the example later

**Reply by lsps9150414** [reply]
Yes, I can understand Chinese.

Btw my goal is described in this post: https://github.com/software-mansion/react-native-gesture-handler/discussions/2071

**Reply by JasonCareter** [reply]
> Yes, I can understand Chinese.
> 
> Btw my goal is described in this post: #2071

```typescript
import React from "react";
import { View, StyleSheet } from "react-native";
import { Gesture, GestureDetector } from "react-native-gesture-handler";
import { useSharedValue } from "react-native-reanimated";

const CompositedLongPressAndPan = () => {

    const isLongPressed = useSharedValue(false);
    const isPanActivated = useSharedValue(false);
    const longPress = Gesture.LongPress()
        .onStart(e => {
            console.log('long press activated')
            isLongPressed.value = true;
        })
    const pan = Gesture.Pan()
        .manualActivation(true)
        .onTouchesMove((event, state) => {
            if (isLongPressed.value && !isPanActivated.value) {
                console.log('pan gesture activated');
                isPanActivated.value = true;
                state.activate();
            }
        })
        .onUpdate((event) => {
            let { translationX, translationY } = event;
            console.log(`transx => ${translationX}  | transy => ${translationY}`);
        })
        .onEnd((event) => {
            isLongPressed.value = false;
            isPanActivated.value = false;
        })
    return (
        <View style={styles.page}>
            <GestureDetector gesture={Gesture.Simultaneous(longPress, pan)}>
                <View style={styles.target} />
            </GestureDetector>
        </View>
    )
}
export default CompositedLongPressAndPan;

const styles = StyleSheet.create({
    target: {
        height: 200,
        width: 200,
        backgroundColor: '#f2f2f3'
    },
    page: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center'
    }
})
```

**Reply by JasonCareter** [reply]
> Yes, I can understand Chinese.
> 
> Btw my goal is described in this post: #2071

if you're doing something like iPhone's Photo App, maybe Image Viewer ? I used to written a component something like that , it might help you with your question related to #2071  . you can just copy the source code of this file [ImageViewer.tsx](https://github.com/yaaliuzhipeng/react-native-tornador-ui/blob/main/components/images/ImageViewer.tsx) and test it

**Reply by lsps9150414** [reply]
@yaaliuzhipeng this is very helpful, thank you!

**Reply by JasonCareter** [reply]
> @yaaliuzhipeng this is very helpful, thank you!

u're welcome 😄 since it did helped you ,maybe you can post it to the discussion mentioned above so that someone else who run into the same problem can also get help from it .

**Reply by Bayramito** [reply]
Hi, i am looking a similar solution  but with tap gesture combined with long press and pan.
In this case how could be the solution @yaaliuzhipeng  ? Thank you

**Reply by JasonCareter** [reply]
do you mean tap、long press and pan, three gestures combination?

**Reply by Bayramito** [reply]
yes,  do you know about whatsapp's voice recording button? 

it has 3 gestures.
onTap => fires onPress event = Gestures.Tap
onHold => stars recording =LongPress with Pan ( to be able to swipe)
onEnd => stops recording =LongPress with Pan again ( to be able to release hold and send element to its origin)
and the cancellation => (which is tricky part) when you swipe x distance from start point, all the gestures cancelling and the button returns to its position..

this is what i want to achieve

**Reply by JasonCareter** [reply]
well, i'm in china right now, people here are not able to use whatsapp under law. perhaps you can post a demonstration video here. wechat (a chat app in china) might be similar to that, in my view, maybe onTap & Pan can give you everything you need in this case.

**Reply by Bayramito** [reply]
There you go,

So,
first of all on onPress event it's showing a bubble  as you can see.
Then when i start holding, it's starting to record... if i release it will send the voice message.
but instead, i swipe to the left smoothly, and after like 100px it cancels everything. (i am not releasing my finger!!!)

https://github.com/software-mansion/react-native-gesture-handler/assets/44513402/ea4b428a-a9f2-4ede-8a3b-53e5623fd465

**Reply by JasonCareter** [reply]
well, it seems one gesture is enough

**Reply by JasonCareter** [reply]
okay, i'm wrong about that , we need  long-press and pan-gesture. first , you should use Simultaneous to let them response to user's interaction at the same time , then you need to record the timestamp when user triggered the event [onTouchDown] of PanGesture, in the [onTouchUp] event, you can calculate the time diff, if it's less than 100 (50 or 200 、depends on you、any num you want) milliseconds and the longPress is not triggered、that's when you should invoke your *onTap* function. what's next, when your long-press triggered, you're supposed to "run some animation and start recording ". in the [onTouchMove] event , you can observer the distance among Axis-X , when it is greater than some value, you are supposed to invoke you *Cancel* function under that condition. well, that's all you need to achieve the effect you want 🎉

**Reply by Bayramito** [reply]
Well, there is a `minDuration` prop to give a delay to long press  :)
but i get your point thanks a lot :)

**Reply by JasonCareter** [reply]
u'are welcome , i've been working on a great project based on flutter for about one year and i didn't write react-native much, so i can't remember much api details of this library 😂
